syntax = "proto3";
package problem;

option optimize_for = SPEED;

message Matrix {
  repeated float time = 1 [packed=true];
  repeated float distance = 2 [ packed = true ];
}

message Location{
  double longitude = 1;
  double latitude = 2;
  uint32 matrix_index = 3;
}

message Service {
  repeated float quantities = 1;
  float duration = 2;
  uint32 matrix_index = 3;
  repeated uint32 compatible_vehicle_indices = 4;
  repeated string skills = 5;
  repeated int32 day_indices = 6;
  Location location = 7;
  string name = 9;
}

message Capacity {
  float limit = 1;
}

message Vehicle {
  string name = 1;
  repeated Capacity capacities = 2;
  Location start_location = 4;
  Location end_location = 5;
  double duration = 6;
  repeated int32 day_indices = 7; //skill_OR
  uint32 matrix_index = 8;
  repeated string skills = 9; //skill_AND
  repeated float load = 11;
  double metric_limit = 12;
  Location centroid = 13;
  repeated uint32 assigned_service_indices = 14;

}
message Solution {
  repeated int32 assignment = 1;
}
message Options{
  uint32 max_iteration = 1;

}
message Problem {
  repeated Vehicle vehicles = 1;
  repeated Service services = 2;
  repeated Matrix matrices = 3;
  Solution solutions = 4;
  repeated string unit_labels = 5;
  Options options = 6;
  double cumulated_metric = 7;
}
